### PR TITLE
Make blockRequestUntil work with MaxSimultaneousRequests and handle e…

### PR DESCRIPTION
**Pull request by mariarek**
…rrors

### CODE REVIEW

**Comment by mariarek** [comment]
Also fixes #32

File: test/SimpleWebRequest.spec.ts [hunk 1]
```diff
@@ -80,6 +93,7 @@ describe('SimpleWebRequest', () => {
         expect(
             () => new SimpleWebRequest<string>(url, method, {}, () => headers).start()
         ).toThrowError(`Don't set Accept with options.headers -- use it with the options.acceptType property`)
+        test_resetQueues();
```
Comment (berickson1) [inline test/SimpleWebRequest.spec.ts]: why is this needed? Shoudln't we clear the queue by executing as a part of the tests?

Comment (mariarek) [reply test/SimpleWebRequest.spec.ts]: If it throws an error, the queue is not cleared. Added it since my other tests were failing, as they test the queue order :) 
Not sure if this should be addressed in the throwing code itself - WDYT?

File: test/SimpleWebRequest.spec.ts [hunk 2]
```diff
@@ -92,7 +106,158 @@ describe('SimpleWebRequest', () => {
...
         ).toThrowError(`Don't set Content-Type with options.headers -- use it with the options.contentType property`)
+        test_resetQueues();
     })
 
+    describe('blocking', () => {
+        let maxRequests = 0;
+
+        beforeEach(() => {
+            maxRequests = SimpleWebRequestOptions.MaxSimultaneousRequests;
+            SimpleWebRequestOptions.MaxSimultaneousRequests = 0;
```
Comment (ms-markda) [inline test/SimpleWebRequest.spec.ts]: [Same other comment, you can spy instead of using your own variable.]

File: test/SimpleWebRequest.spec.ts [hunk 3]
```diff
@@ -92,7 +106,158 @@ describe('SimpleWebRequest', () => {
...
+        it('executes the requests by priority and age', () => {
+            const url = faker.internet.url();
+            const method = 'GET';
+            const onSuccess1 = jasmine.createSpy('onSuccess1');
+            const onSuccess2 = jasmine.createSpy('onSuccess2');
+            const onSuccess3 = jasmine.createSpy('onSuccess3');
+            const onSuccess4 = jasmine.createSpy('onSuccess4');
+            const status = 200;
+    
+            new SimpleWebRequest<string>(url, method, { priority: WebRequestPriority.Low }).start().then(onSuccess1);
```
Comment (ms-markda) [inline test/SimpleWebRequest.spec.ts]: Suggestion: Use `onSuccessN` in the expected order, not the created order. It is harder to follow that you expect 2->4->1->3.

File: test/SimpleWebRequest.spec.ts [hunk 4]
```diff
@@ -92,7 +106,158 @@ describe('SimpleWebRequest', () => {
...
+    
+        it('after the request is unblocked, it\'s returned to the queue with correct priority', () => {
+            const url = faker.internet.url();
+            const method = 'GET';
+            const blockDefer = SyncTasks.Defer<void>();
+            const onSuccess1 = jasmine.createSpy('onSuccess1');
+            const onSuccess2 = jasmine.createSpy('onSuccess2');
+            const onSuccess3 = jasmine.createSpy('onSuccess3');
+
+            new SimpleWebRequest<string>(url, method, { priority: WebRequestPriority.High }, undefined, () => blockDefer.promise()).start().then(onSuccess1);
```
Comment (ms-markda) [inline test/SimpleWebRequest.spec.ts]: [Same `onSuccessN` comment as above.]

File: test/SimpleWebRequest.spec.ts [hunk 5]
```diff
@@ -92,7 +106,158 @@ describe('SimpleWebRequest', () => {
...
+            const method = 'GET';
+            const blockDefer = SyncTasks.Defer<void>();
+            const onSuccess1 = jasmine.createSpy('onSuccess1');
+            new SimpleWebRequest<string>(url, method, { priority: WebRequestPriority.Critical }, undefined, () => blockDefer.promise()).start().then(onSuccess1).catch(err => {
+                expect(1).toBe(1);
+                done();
+            });
+            
+            blockDefer.reject('Terrible error');
+
```
Comment (ms-markda) [inline test/SimpleWebRequest.spec.ts]: Nit: Trailing newline.

File: test/SimpleWebRequest.spec.ts [hunk 6]
```diff
@@ -92,7 +106,158 @@ describe('SimpleWebRequest', () => {
...
+        });
+
+        it('fails the request, if the blocking promise rejects', done => {
+            SimpleWebRequestOptions.MaxSimultaneousRequests = 1;
+            const url = faker.internet.url();
+            const method = 'GET';
+            const blockDefer = SyncTasks.Defer<void>();
+            const onSuccess1 = jasmine.createSpy('onSuccess1');
+            new SimpleWebRequest<string>(url, method, { priority: WebRequestPriority.Critical }, undefined, () => blockDefer.promise()).start().then(onSuccess1).catch(err => {
+                expect(1).toBe(1);
```
Comment (ms-markda) [inline test/SimpleWebRequest.spec.ts]: Suggestion: Expect `onSuccess1` to have not been called, instead of `1 to be 1`.

File: src/GenericRestClient.ts [hunk 7]
```diff
@@ -51,12 +51,7 @@ export class GenericRestClient {
             options.sendData = objToPost;
         }
 
-        const promise = this._blockRequestUntil(options);
-        if (!promise) {
-            return this._performApiCallInternal(apiPath, action, options);
-        }
-
-        return promise.then(() => this._performApiCallInternal(apiPath, action, options));
+        return this._performApiCallInternal(apiPath, action, options);
```
Comment (ms-markda) [inline src/GenericRestClient.ts]: You don't need a separate `_performApiCallInternal` any more. Merge with `_performApiCall`.

File: package.json [hunk 8]
```diff
@@ -9,7 +9,8 @@
     "build": "npm run tslint && tsc",
     "tslint": "tslint --project tsconfig.json -r tslint.json --fix || true",
     "test": "npm run clean && karma start --singleRun",
-    "test:watch": "npm run clean && karma start"
+    "test:watch": "npm run clean && karma start",
+    "test:browser": "npm run clean && karma start --browsers=Chrome --single-run=false --auto-watch"
```
Comment (ms-markda) [inline package.json]: This is a breaking change. Version should bump; perhaps a minor bump to `0.2.0` since major is `0`. Should also document breaking changes. Perhaps in the README, or in some other document.

File: test/SimpleWebRequest.spec.ts (hunk 6 above)
Comment (mariarek) [reply test/SimpleWebRequest.spec.ts]: Totally forgot about this :)

File: package.json (hunk 8 above)
Comment (mariarek) [reply package.json]: hows is that a breaking change? I'm just adding a new command.

Comment (ms-markda) [reply package.json]: The entire PR is a breaking change, not just this line.

File: test/SimpleWebRequest.spec.ts (hunk 1 above)
Comment (berickson1) [reply test/SimpleWebRequest.spec.ts]: Oh, that's interesting - not sure what we should do in that case to be honest

File: test/SimpleWebRequest.spec.ts [hunk 9]
```diff
@@ -1,10 +1,21 @@
...
 
 describe('SimpleWebRequest', () => {
-    beforeEach(() => jasmine.Ajax.install());
-    afterEach(() => jasmine.Ajax.uninstall());
+    let catchExceptions = false;
+    const status = 200;
+
+    beforeEach(() => {
+        catchExceptions = SyncTasks.config.catchExceptions;
+        SyncTasks.config.catchExceptions = false;
```
[the remark this answers is not in the document]
Comment (mariarek) [reply test/SimpleWebRequest.spec.ts]: no way to spy on property with no getters

File: package.json (hunk 8 above)
Comment (mariarek) [reply package.json]: updated the version. 
The parts I'm changing were not a part of README before, should I add a section about blocking? Or just add a changelog file?

File: test/SimpleWebRequest.spec.ts [hunk 10]
```diff
@@ -78,8 +92,9 @@ describe('SimpleWebRequest', () => {
         const url = faker.internet.url();
 
         expect(
-            () => new SimpleWebRequest<string>(url, method, {}, () => headers).start()
-        ).toThrowError(`Don't set Accept with options.headers -- use it with the options.acceptType property`)
+            () =>  new SimpleWebRequest<string>(url, method, {}, () => headers).start()
```
Comment (ms-markda) [inline test/SimpleWebRequest.spec.ts]: Nit: Remove extra space between `=>` and `new`.

File: src/SimpleWebRequest.ts [hunk 11]
```diff
@@ -213,8 +215,35 @@ export abstract class SimpleWebRequestBase<TOptions extends WebRequestOptions =
     protected static checkQueueProcessing() {
         while (requestQueue.length > 0 && executingList.length < SimpleWebRequestOptions.MaxSimultaneousRequests) {
             const req = requestQueue.shift()!!!;
-            executingList.push(req);
-            req._fire();
+            const blockPromise = (req._blockRequestUntil&& req._blockRequestUntil()) || SyncTasks.Resolved();
```
Comment (ms-markda) [inline src/SimpleWebRequest.ts]: Nit: Add space between `_blockRequestUntil` and `&&`.

File: src/SimpleWebRequest.ts [hunk 12]
```diff
@@ -626,7 +656,7 @@ export class SimpleWebRequest<TBody, TOptions extends WebRequestOptions = WebReq
                 // Some browsers error when you try to read status off aborted requests
             }
         } else {
-            statusText = 'Browser Error - Possible CORS or Connectivity Issue';
+            statusText = errorStatusText || 'Browser Error - Possible CORS or Connectivity Issue';
```
Comment (ms-markda) [inline src/SimpleWebRequest.ts]: Should the above usage of `errorStatusText` also come first? `this._xhr.statusText || errorStatusText` -> `errorStatusText || this._xhr.statusText` ?

File: src/SimpleWebRequest.ts [hunk 13]
```diff
@@ -213,8 +215,35 @@ export abstract class SimpleWebRequestBase<TOptions extends WebRequestOptions =
...
+    protected _removeFromQueue(): void {
+        // Pull it out of whichever queue it's sitting in
+        if (this._xhr) {
+            _.pull(executingList, this);
+        } else {
+            _.pull(requestQueue, this);
+        }
+    }
+
+    protected _assertAndClean(expression: any, message: string): void {
```
Comment (ms-markda) [inline src/SimpleWebRequest.ts]: `expression: boolean` instead of `any`. The caller can force a boolean as needed.

File: src/SimpleWebRequest.ts [hunk 14]
```diff
@@ -240,7 +269,7 @@ export abstract class SimpleWebRequestBase<TOptions extends WebRequestOptions =
             const timeoutSupported = timeoutSupportStatus;
              // Use manual timer if we don't know about timeout support
             if (timeoutSupported !== FeatureSupportStatus.Supported) {
-                assert.ok(!this._requestTimeoutTimer, 'Double-fired requestTimeoutTimer');
+                this._assertAndClean(!this._requestTimeoutTimer, 'Double-fired requestTimeoutTimer');
                 this._requestTimeoutTimer = SimpleWebRequestOptions.setTimeout(() => {
```
Comment (ms-markda) [inline src/SimpleWebRequest.ts]: Why clean and then set a timer for `abort`? If this cleans then should you skip the timer?

File: src/SimpleWebRequest.ts [hunk 15]
```diff
@@ -262,7 +291,7 @@ export abstract class SimpleWebRequestBase<TOptions extends WebRequestOptions =
                     this._timedOut = true;
                     // Set aborted flag to match simple timer approach, which aborts the request and results in an _respond call
                     this._aborted = true;
-                    this._respond();
+                    this._respond('Aborted');
```
Comment (ms-markda) [inline src/SimpleWebRequest.ts]: Question: Should this be `'Timeout'` instead?

File: src/SimpleWebRequest.ts [hunk 16]
```diff
@@ -344,14 +373,14 @@ export abstract class SimpleWebRequestBase<TOptions extends WebRequestOptions =
...
+                this._assertAndClean(false, 'Don\'t set Content-Type with options.headers -- use it with the options.contentType property');
                 return;
             }
             if (headerLower === 'accept') {
-                assert.ok(false, 'Don\'t set Accept with options.headers -- use it with the options.acceptType property');
+                this._assertAndClean(false, 'Don\'t set Accept with options.headers -- use it with the options.acceptType property');
                 return;
             }
-            assert.ok(!headersCheck[headerLower], 'Setting duplicate header key: ' + headersCheck[headerLower] + ' and ' + key);
+            this._assertAndClean(!headersCheck[headerLower], 'Setting duplicate header key: ' + headersCheck[headerLower] + ' and ' + key);
```
Comment (ms-markda) [inline src/SimpleWebRequest.ts]: If any of these clean then you should prevent the `xhr` from sending as well. [Note: These `return` are inside `_.forEach`]

File: test/SimpleWebRequest.spec.ts [hunk 17]
```diff
@@ -178,84 +179,84 @@ describe('SimpleWebRequest', () => {
...
             const url = faker.internet.url();
             const method = 'GET';
             const blockDefer = SyncTasks.Defer<void>();
-            const onSuccess1 = jasmine.createSpy('onSuccess1');
-            new SimpleWebRequest<string>(url, method, { priority: WebRequestPriority.Critical }, undefined, () => blockDefer.promise()).start().then(onSuccess1).catch(err => {
-                expect(1).toBe(1);
+            new SimpleWebRequest<string>(url, method, { priority: WebRequestPriority.Critical }, undefined, () => blockDefer.promise()).start()
+            .then(() => fail(), (err: WebErrorResponse) => {
+                expect(err.statusCode).toBe(0);
+                expect(err.statusText).toBe('Error in _blockRequestUntil: Terrible error');
```
Comment (ms-markda) [inline test/SimpleWebRequest.spec.ts]: I don't think you should have "_blockRequestUntil" in the `statusText`. That property feels too "public API"-like, but a method name feels too "internal API"-like. Consider rewording.

File: src/SimpleWebRequest.ts (hunk 12 above)
Comment (mariarek) [reply src/SimpleWebRequest.ts]: No - if we have a real status text - use that instead of the fallback.
But if we don't have the xhr at all - first use the provided value and then the fallback 
Basically priority is :
real status > passed status > generic 'Browser Error (...)' constant

File: src/SimpleWebRequest.ts (hunk 13 above)
Comment (mariarek) [reply src/SimpleWebRequest.ts]: assert.ok takes any, so I kept it consistent

File: src/SimpleWebRequest.ts (hunk 16 above)
Comment (mariarek) [reply src/SimpleWebRequest.ts]: The assert prevents that from happening - I want to leave the library in a good state if somebody decides to catch the exception.
If they do -> subsequent requests work
If they don't -> there are no subsequent requests
Returning here only makes sense if someone strips/replaces the asserts - in which case they should be responsible for making sure the library still works

File: test/SimpleWebRequest.spec.ts (hunk 17 above)
Comment (mariarek) [reply test/SimpleWebRequest.spec.ts]: How would you name it? It can only happen if the caller has overwritten the `_blockRequestUntil` method, in which case it feels appropriate to point out that this is what's causing the failure

File: src/SimpleWebRequest.ts [hunk 18]
```diff
@@ -213,8 +215,35 @@ export abstract class SimpleWebRequestBase<TOptions extends WebRequestOptions =
...
+            blockPromise.then(() => {
+                if (executingList.length < SimpleWebRequestOptions.MaxSimultaneousRequests) {
+                    executingList.push(req);
+                    req._fire();
+                } else {
+                    req._enqueue();
+                }
+            }, err => {
+                // fail the request if the block promise is rejected
+                req._respond('Error in _blockRequestUntil: ' + err.toString());
```
Comment (ms-markda) [inline src/SimpleWebRequest.ts]: Nit: Remove unnecessary `.toString()`.

Comment (ms-markda) [inline src/SimpleWebRequest.ts]: Suggestion: `"Rejected at end of blocking: " + err`. [1] "Rejection" is not the same terminology as "Error", and [2] removes reference to internal API.

File: test/SimpleWebRequest.spec.ts [hunk 19]
```diff
@@ -88,11 +103,160 @@ describe('SimpleWebRequest', () => {
...
+            const method = 'GET';
+            const blockDefer = SyncTasks.Defer<void>();
+            new SimpleWebRequest<string>(url, method, { priority: WebRequestPriority.Critical }, undefined, () => blockDefer.promise()).start()
+            .then(() => fail(), (err: WebErrorResponse) => {
+                expect(err.statusCode).toBe(0);
+                expect(err.statusText).toBe('Error in _blockRequestUntil: Terrible error');
+                done();
+            });
+            
+            blockDefer.reject('Terrible error');
```
Comment (ms-markda) [inline test/SimpleWebRequest.spec.ts]: Suggestion: Put `'Terrible error'` in a constant since it is used in two places. The point of the test is for this and that to be the same; the exact string does not matter.
